src/FLOWVPM_timeintegration.jl

```diff
@@ -256,89 +270,104 @@ Steps the field forward in time by dt in a third-order low-storage Runge-Kutta
 integration scheme using the VPM reformulation. See Notebook entry 20180105
 (RK integration) and notebook 20210104 (reformulation).
 """
-function rungekutta3(pfield::ParticleField{R, <:ReformulatedVPM{R2}, V, <:SubFilterScale},
-                     dt::Real; relax::Bool=false ) where {R, V, R2}
+function rungekutta3(pfield::ParticleField{R, <:ReformulatedVPM{R2}, V, <:SubFilterScale, <:Any, <:Any, <:Any},
+                     dt::R3; relax::Bool=false, custom_UJ=nothing) where {R, V, R2, R3}
 
-    # Storage terms: qU <=> p.M[:, 1], qstr <=> p.M[:, 2], qsmg2 <=> p.M[1, 3],
-    #                      qsmg <=> p.M[2, 3], Z <=> MM[4], S <=> MM[1:3]
+    # Storage terms: qU <=> p.M[:, 1], qstr <=> p.M[:, 2], qsmg2 <=> get_M(p)[7],
+    #                      qsmg <=> get_M(p)[8], Z <=> MM[4], S <=> MM[1:3]
 
     # Calculate freestream
-    Uinf::Array{<:Real, 1} = pfield.Uinf(pfield.t)
-
-    MM::Array{<:Real, 1} = pfield.M
-    f::R2, g::R2 = pfield.formulation.f, pfield.formulation.g
-    zeta0::R = pfield.kernel.zeta(0)
+    Uinf::Array{R, 1} = R.(pfield.Uinf(pfield.t)) # now infers its type from pfield. although tbh this isn't correct; a functor for U would be a cleaner implementation.
 
+    MM::Array{R, 1} = pfield.M # eltype(pfield.M) = R
+    f::R2, g::R2 = pfield.formulation.f, pfield.formulation.g # formulation floating-point type may end up as Float64 even if AD is used. (double check this)
```

Comment:
It should be. It's probably safest to leave it them with their own type (i.e., keep the incoming changes), since f and g are basically constants tied to the formulation (as opposed to the type of the particle states).